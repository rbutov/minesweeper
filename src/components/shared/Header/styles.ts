import styled from 'styled-components';
import { TimerContainer } from 'components/shared/Timer/styles';
import { CounterContainer } from 'components/shared/Counter/styles';
import { FaceContainer } from 'components/shared/Face/styles';

const HeaderContainer = styled.div`
  flex-grow: 1;
  height: 48px;
  background: silver;
  position: relative;

  ${TimerContainer} {
    position: absolute;
    right: 2px;
    top: 2px;
  }
  ${CounterContainer} {
    position: absolute;
    left: 2px;
    top: 2px;
  }
  ${FaceContainer} {
    position: relative;
    top: 4px;
    margin: auto;
  }
`;

export { HeaderContainer };
