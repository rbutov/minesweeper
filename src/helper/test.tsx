import React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';
import { ReactWrapper, ShallowWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { confStore } from 'store';

type Wrapper<P = unknown, S = unknown> = ShallowWrapper<P, S> | ReactWrapper<P, S>;
type MountFnType<P, S> = (
  node: React.ReactElement<P>,
  options?: Record<string, unknown>
) => Wrapper<P, S>;

const wrapAppComponent = <P, S>(
  renderer: MountFnType<P, S>,
  element: JSX.Element,
  other?: {
    store?: Store;
  }
): Wrapper<P, S> => {
  const coreStore = confStore();

  return renderer(<Provider store={other?.store ? other.store : coreStore}>{element}</Provider>);
};

const waitForComponentToPaint = async <P, S>(
  wrapper: Wrapper<P, S>,
  other?: { delay?: number; props?: Pick<P, keyof P> }
) => {
  await act(async () => {
    await runAllPromises();
    await new Promise((resolve) => setTimeout(resolve, other?.delay ? other.delay : 0));
    wrapper.setProps({ ...other?.props });
    wrapper.update();
  });
};

const runAllPromises = async () => {
  return new Promise<void>((resolve) => {
    setTimeout(resolve, 0);
  });
};

export { wrapAppComponent, waitForComponentToPaint };
export type { Wrapper };
