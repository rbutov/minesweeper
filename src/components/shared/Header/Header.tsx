import React from 'react';
import { getFace } from 'helper/game';
import { useAppSelector } from 'hooks/useAppSelector';
import { Face } from 'components/shared/Face';
import { Counter } from 'components/shared/Counter';
import { Timer } from 'components/shared/Timer';
import { HeaderContainer } from './styles';

const Header = () => {
  const { size, revealedKeys, failedMineKey, minesCount } = useAppSelector((state) => state.field);

  return (
    <HeaderContainer data-testid={'header'}>
      <Counter />
      <Face type={getFace({ size, revealedKeys, failedMineKey, minesCount })} />
      <Timer />
    </HeaderContainer>
  );
};

export { Header };
