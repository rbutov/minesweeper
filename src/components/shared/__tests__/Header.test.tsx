import React from 'react';
import { wrapAppComponent } from 'helpers/test';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Header } from '../Header';

describe('<Header />', () => {
  it('_render', () => {
    const wrapper = wrapAppComponent(mount, <Header />);

    const timer = wrapper.find({ 'data-testid': 'timer' }).first().getElement();
    expect(wrapper.contains(timer)).toBeTruthy();
    const counter = wrapper.find({ 'data-testid': 'counter' }).first().getElement();
    expect(wrapper.contains(counter)).toBeTruthy();
    const face = wrapper.find({ 'data-testid': 'face' }).first().getElement();
    expect(wrapper.contains(face)).toBeTruthy();

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
