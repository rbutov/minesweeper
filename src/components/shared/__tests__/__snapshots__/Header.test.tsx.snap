// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> _render 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Header>
    <styled.div
      data-testid="header"
    >
      <div
        className="sc-jRQBWg SDskj"
        data-testid="header"
      >
        <Counter>
          <styled.div
            data-testid="counter"
          >
            <div
              className="sc-hKwDye cHUlqo"
              data-testid="counter"
            >
              <Number
                num={0}
              >
                <styled.div
                  data-testid="number-0"
                  num={0}
                >
                  <div
                    className="sc-dkPtRN jWYRyZ"
                    data-testid="number-0"
                  />
                </styled.div>
              </Number>
              <Number
                num={1}
              >
                <styled.div
                  data-testid="number-1"
                  num={1}
                >
                  <div
                    className="sc-dkPtRN kQGVMA"
                    data-testid="number-1"
                  />
                </styled.div>
              </Number>
              <Number
                num={0}
              >
                <styled.div
                  data-testid="number-0"
                  num={0}
                >
                  <div
                    className="sc-dkPtRN jWYRyZ"
                    data-testid="number-0"
                  />
                </styled.div>
              </Number>
            </div>
          </styled.div>
        </Counter>
        <Face
          type="unpressed"
        >
          <styled.div
            data-testid="face"
            onClick={[Function]}
            type="unpressed"
          >
            <div
              className="sc-gsDKAQ hZfqEV"
              data-testid="face"
              onClick={[Function]}
              type="unpressed"
            />
          </styled.div>
        </Face>
        <Timer>
          <styled.div
            data-testid="timer"
          >
            <div
              className="sc-eCImPb nFXVg"
              data-testid="timer"
            >
              <Number
                num={0}
              >
                <styled.div
                  data-testid="number-0"
                  num={0}
                >
                  <div
                    className="sc-dkPtRN jWYRyZ"
                    data-testid="number-0"
                  />
                </styled.div>
              </Number>
              <Number
                num={0}
              >
                <styled.div
                  data-testid="number-0"
                  num={0}
                >
                  <div
                    className="sc-dkPtRN jWYRyZ"
                    data-testid="number-0"
                  />
                </styled.div>
              </Number>
              <Number
                num={0}
              >
                <styled.div
                  data-testid="number-0"
                  num={0}
                >
                  <div
                    className="sc-dkPtRN jWYRyZ"
                    data-testid="number-0"
                  />
                </styled.div>
              </Number>
            </div>
          </styled.div>
        </Timer>
      </div>
    </styled.div>
  </Header>
</Provider>
`;
